### PR TITLE
Add ChargeArray

### PR DESCRIPTION
**Pull request by mganahl**
Base class for BlockSparseTensor

### CODE REVIEW

File: tensornetwork/block_sparse/block_tensor.py [hunk 1]
```diff
@@ -626,3 +627,256 @@ def _find_transposed_diagonal_sparse_blocks(
...
+      Tuple: A tuple of `Index` objects.
+    """
+
+    indices = [
+        Index([self._charges[n]
+               for n in s], [self._flows[n]
+                             for n in s])
+        for s in self._order
+    ]
+    return tuple(indices)
```
Comment (Thenerdstation) [inline tensornetwork/block_sparse/block_tensor.py]: Yikes triple nesting.

File: tensornetwork/block_sparse/block_tensor.py [hunk 2]
```diff
@@ -626,3 +627,256 @@ def _find_transposed_diagonal_sparse_blocks(
...
+        new_shape.append(s)
+
+    # a few simple checks
+    if np.prod(new_shape) != np.prod(self.shape):
+      raise ValueError("A tensor with {} elements cannot be "
+                       "reshaped into a tensor with {} elements".format(
+                           np.prod(self.shape), np.prod(new_shape)))
+
+    flat_dims = flatten(
+        [[self._charges[n].dim for n in o] for o in self._order])
```
Comment (Thenerdstation) [inline tensornetwork/block_sparse/block_tensor.py]: I believe you can do `[self._charges[n].dim for n in o for o in self._order]`

Comment (mganahl) [reply tensornetwork/block_sparse/block_tensor.py]: you're right, that's  much better (you need to change order of the loops though)

File: tensornetwork/block_sparse/block_tensor.py (hunk 1 above)
Comment (mganahl) [reply tensornetwork/block_sparse/block_tensor.py]: fixed that one

**Comment by mganahl** [comment]
Hey @Thenerdstation, fixed your comments and made some additional changes.

**Pull request merged**